addons/libs/dialog.lua

```diff
@@ -0,0 +1,158 @@
+-- This library was written to help find the ID of a known
+-- action message corresponding to an entry in the dialog tables.
+-- While the IDs can be collected in-game, they occasionally
+-- change and would otherwise need to be manually updated.
+-- It can also be used to find and decode an entry given the ID.
+
+-- Common parameters:
+-- dat: The entire content of the zone dialog DAT file
+-- i.e. local dat = io.open('path/to/dialog/DAT', 'rb'):read('*a')
+-- entry: The string you are looking for. If you do not know the
+-- entire string, use dev_find_substring.
+
+local xor = require('bit').bxor
+require('pack')
+local string = require('string')
+local unpack = string.unpack
+local pack = string.pack
+local find = string.find
+
+local dialog = {}
+
+local function decode(int)
+    return xor(int, 0x80808080)
+end
+local encode = decode
+
+local floor = require('math').floor
+local function binary_search(pos, dat, n)
+    local l, r, m = 1, n
+    while l < r do
+        m = floor((l+r)/2)
+        if decode(unpack('<I', dat, 1+4*m)) < pos then
+        -- offset given by mth ID < offset to string
+            l = m + 1
+        else
+            r = m
+        end
+    end
+    return l-2 -- we want the index to the left of where "pos" would be placed
+end
+
+local function plain_text_gmatch(text, substring, n)
+    n = n or 1
+    return function()
+        local pos = find(text, substring, n, true)
+        if pos then n = pos + 1 end
+        return pos
+    end
+end
+
+function dialog.dev_get_offset(dat, id) -- sanity check function
+    return string.format('%x', decode(unpack('<I', dat, 5+4*id)))
```

Comment:
Maybe a nitpick, but you added locals for the other string functions, may as well do it for `string.format` :D Same below for `string.sub`, `string.gsub`, `string.char`, `string.byte`.